File: src/includes.go
package main

/*
	object to call in yaml files
*/
import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

type ObjectLog interface {
	exec() string
	init(a *Action)
}

// ###############
// Display packages name and version if installed
// ###############

type PkgVer struct {
	pkgs string
}

func (p *PkgVer) init(a *Action) {
	p.pkgs = strings.ToLower(a.Pkgs)
}

func (p PkgVer) exec() string {
	cmd := fmt.Sprintf("LANG=C pacman -Qi %s |awk -F':' '/^Name/ {{n=$2}} /^Ver/ {{print n\": \"$2}}'", p.pkgs)
	if p.pkgs != "" {
		out, err := exec.Command("bash", "-c", cmd).Output()
		if err == nil {
			return string(out)
		}
	}
	return ""
}

// ###############
// Display journald log but with error level
// ###############

type JournalType struct {
	UID               string `json:"_UID"`
	Cmdline           string `json:"_CMDLINE,omitempty"`
	SyslogIdentifier  string `json:"SYSLOG_IDENTIFIER"`
	Comm              string `json:"_COMM"`
	RealtimeTimestamp string `json:"__REALTIME_TIMESTAMP"`
	Priority          string `json:"PRIORITY"`
	Message           string `json:"MESSAGE"`
}

type Journald struct {
	level int
	count int
}

func (j *Journald) init(a *Action) {
	if a.Level == 0 {
		a.Level = 3
	}
	if a.Count == 0 {
		a.Count = 32
	}
	j.count = a.Count
	j.level = a.Level
}

func (j Journald) exec() string {
	const f = "__REALTIME_TIMESTAMP,PRIORITY,_COMM,_UID,MESSAGE,_CMDLINE,SYSLOG_IDENTIFIER"
	cmd := fmt.Sprintf("journalctl -b0 -p%d -qr -n%d --no-pager --output-fields=\"%s\" -o json", j.level, j.count, f)
	ret := ""
	out, err := exec.Command("bash", "-c", cmd).Output()
	if err == nil {
		var dat []JournalType
		if err := json.Unmarshal([]byte("["+strings.ReplaceAll(string(out), "}\n{", "},\n{")+"]"), &dat); err != nil {
			panic(err)
		}
		oldentry := ""
		for _, j := range dat {

			i, err := strconv.ParseInt(j.RealtimeTimestamp[0:10], 10, 64)
			if err != nil {
				i = 0
			}
			tm := time.Unix(i, 0)

			/*
				cmdline := j.Cmdline
				if cmdline == "" {
					cmdline = j.SyslogIdentifier
				}
			*/
			entry := fmt.Sprintf("(%s) %s[%s]: %s", j.Priority, j.Comm, j.UID, j.Message)
			// no repeat if same entry
			if entry != oldentry {
				oldentry = entry
				ret = fmt.Sprintf("%s%s %s\n", ret, tm.Format("2006-01-02 15:04:05"), entry)
			}
		}
		return ret
	}
	return ""
}

type LogsActivity struct {
	count int
	regex *regexp.Regexp
}

/*
[2021-10-28T04:01:19+0200] [PACMAN] starting full system upgrade
[2021-10-28T10:11:07+0200] [ALPM] transaction started
[2021-10-31T01:22:07+0200] [ALPM] upgraded xmlsec (1.2.32-1 -> 1.2.33-1)
[2021-08-06T21:08:06+0200] [ALPM] removed tk (8.6.11.1-1)
[2021-08-05T21:05:23+0200] [ALPM] installed libtg_owt (0.git6.91d836d-2)
[2021-10-30T15:30:22+0200] [ALPM] transaction completed
*/

func (l *LogsActivity) init(a *Action) {
	if a.Count == 0 {
		a.Count = 30
	}
	if a.Regex == "" {
		a.Regex = ".*"
	}
	l.count = a.Count
	l.regex = regexp.MustCompile(a.Regex)
}

func (l LogsActivity) exec() (ret string) {
	now := time.Now().AddDate(0, -0, -l.count)

	file, err := os.Open("/var/log/pacman.log")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	calendar := make(map[string]int)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) > 20 && line[0] == '[' {
			d := line[1:11]
			t, _ := time.Parse("2006-01-02", d)
			if t.After(now) {
				if strings.Index(line, "[ALPM] ") == -1 {
					continue
				}
				//run regex ...
				if l.regex.MatchString(line) {
					if _, ok := calendar[d]; ok {
						calendar[d] += 1
					} else {
						calendar[d] = 1
					}
				}
			}
		}
	}
	max := func() (max int) {
		for _, v := range calendar {
			if v > max {
				max = v
			}
		}
		return max
	}()
	sortc := func(map[string]int) []string {
		keys := make([]string, len(calendar))
		i := 0
		for k := range calendar {
			keys[i] = k
			i++
		}
		sort.Strings(keys)
		return keys
	}
	for _, i := range sortc(calendar) {
		c := calendar[i]
		pourcent := int(math.Round((float64(c) * float64(100)) / float64(max)))
		ret += fmt.Sprintf("%s %v%-3v%v %s\n", i, COLOR_GREEN, c, COLOR_NONE, strings.Repeat("━", pourcent))
	}
	return ret
}

File: zip.py/src/includes.py
import subprocess
from datetime import datetime
import json

class IncludeJob:
    """ Abstract class for call include object from yaml """

    def __init__(self, action: dict) -> None:
        self.action = action

    def __call__(self):
        raise RuntimeError(f"Abstract Class 🥵")

    def __str__(self) -> str:
        raise RuntimeError(f"Abstract Class 🥵")



class PkgVer(IncludeJob):
    """ list packages with version """

    def __init__(self, action) -> None:
        super().__init__(action)
        self.pkgs = action['pkgs'].lower().split() if action.get('pkgs') else []
        self.logs = ""

    def __call__(self):
        if not self.pkgs:
            return self
        cmd = f"pacman -Qi {' '.join(self.pkgs)} |awk -F':' '/^Name/ {{n=$2}} /^Ver/ {{print n\": \"$2}}'"
        with subprocess.Popen(
                [f"env TERM=xterm LANG=C {cmd}"],
                universal_newlines=True,
                stdout=subprocess.PIPE,
                shell=True,
                text=True
            ) as process:
                self.logs = process.stdout.read()
        return self

    def __str__(self) -> str:
        return self.logs


class Journald(IncludeJob):
    def __init__(self, action) -> None:
        super().__init__(action)
        self.logs = ()
        self.level = action['level'] if action.get('level') else 3

    def _print_item(self, item, old_date) -> tuple[str, str]:
        d = item['DATE'][:16]
        if d == old_date:
            d = ""
        return item['DATE'][:16], f"{d}\n\t({item['PRIORITY']}) [{item['_UID']:>4}] {item['_CMDLINE']}\n\t{item['MESSAGE']}"

    def __str__(self) -> str:
        max = 44
        i = 1
        ret = ""
        old_date = ""
        for item in self.logs:
            i += 1
            old_date, outstr = self._print_item(item, old_date)
            ret = f"{ret}\n{outstr}"
            if i > max:
                break
        return f"\n{ret}\n"

    def __call__(self):
        cmd = f'SYSTEMD_COLORS=0 /usr/bin/journalctl -b0 -p{self.level} --no-pager -o json'
        with subprocess.Popen(cmd, stdout=subprocess.PIPE, bufsize=-1, universal_newlines=True, text=True, shell=True) as proc:
            for line in proc.stdout:
                data = json.loads(line)
                item = {key: value for key, value in data.items() if key in ["PRIORITY", "MESSAGE", "_CMDLINE", "_UID", "__REALTIME_TIMESTAMP"]}
                dt_object = datetime.fromtimestamp(int(item["__REALTIME_TIMESTAMP"][0:10]))
                item['DATE'] = str(dt_object)
                if '_UID' not in item.keys():
                    item['_UID'] = "0"
                if '_CMDLINE' not in item.keys():
                    item['_CMDLINE'] = data["SYSLOG_IDENTIFIER"]
                self.logs = self.logs + (item,)
        if proc.returncode != 0:
            exit(proc.returncode)
        return self
    
class PkgExists(IncludeJob):
    """ package installed ? """
    def __init__(self, action) -> None:
        super().__init__(action)
        self.ok = False
        self.pkgs = action.get('pkgs',"*")

    def __str__(self) -> str:
        return f"\ninstalled: {self.pkgs}\n" if self.ok else ""

    def __call__(self):
        try:
            subprocess.check_output(f'pacman -Qq {self.pkgs} >/dev/null 2>&1', shell=True)
            self.ok = True
        except subprocess.CalledProcessError:
            self.ok = False
        return self
